src/common/types.jl

```diff
@@ -4,6 +4,9 @@ using StaticArrays: setindex
 # Deduce the value type (floating-point type for storing plain an static data
 # in Model and PlaneWaveBasis) from e.g. an interval or a dual type.
 value_type(T) = T
+# Lossy: force interpretation of value as a determined type.
+convert_enforced(::Type{T}, x) where {T <: Real} = real(x)
+convert_enforced(::Type{T}, x) where {T <: Complex} = x
```

Comment:
I still wonder about this function and what it is for: Would you not want to do a `convert(T, x)`  as well ? Maybe also stick a `convert_enforced(::Type{T}, x)::T where {T <: xyz} = ... ` to ensure that the type can be propagated by the compiler.

Reply:
> Not sure I understand. convert(::Real, x::Complex) does not make sense as we lose information.

Yes, but what I meant was the following:
```suggestion
convert_enforced(::Type{T}, x)::T where {T <: Real}    = convert(T, real(x))
convert_enforced(::Type{T}, x)::T where {T <: Complex} = convert(T, x)
```

Reply:
does this not do the trick ? 

Basically I'm not doubting the function is needed, I'm mostly trying to understand what it is supposed to do and then trying to find a name that is not misleading. `convert` can deal with vectors and arrays, but then the `T` will be an array type as well for example.

Reply:
Maybe what we want is actually the [Adapt.jl](https://github.com/JuliaGPU/Adapt.jl/) pattern ?

Reply:
Ok then maybe `convert_enforced` is not the right name. For me it does a `convert`, but is happy to do inexact converts, which would otherwise fail with an Inexact error. Is that the intention ?

Reply:
Ok then why does
```
convert_enforced(::Type{T}, x) where {T <: Real}    = convert(T, real(x))::T
convert_enforced(::Type{T}, x)                      = convert(T, x)::T
```
not do everything you want ? It does a normal convert for complex numbers and other types and first strips the imaginary part when converting to a real number.